pkg/engine/planner/logical/key_collector.go

```diff
@@ -0,0 +1,141 @@
+package logical
+
+import (
+	"sort"
+
+	"github.com/grafana/loki/v3/pkg/logql/log"
+	"github.com/grafana/loki/v3/pkg/logql/syntax"
+)
+
+// CollectRequestedKeys walks the AST and collects keys that need to be extracted
+// from logfmt parsing in metric queries, along with numeric type hints for unwrap operations.
+//
+// Returns:
+//   - keys: sorted list of unique keys to extract
+//   - hints: map of field name to type hint ("int64", "float64", or "duration")
+//
+// TODO(twhitney): These nil cases are temporary and will be removed once we implement support for log queries.
+// Returns nil, nil if:
+//   - The query is not a metric query (no range or vector aggregation)
+//   - The query does not use logfmt parsing
+//
+// This function is intentionally limited in scope to support optimization of
+// logfmt metric queries in the v2 engine. Support for log queries and other
+// parsers may be added in the future.
+func CollectRequestedKeys(expr syntax.Expr) ([]string, map[string]string) {
```

Comment:
IMO this is closer to the domain of an optimization phase (like some kind of projection/predicate pushdown) that we do in the physical planner. 

I'm not sure we should do this here, especially since it requires deeper analysis of the entire LogQL expression.

Reply:
+1 `pkg/engine/planner/physical/optimizer.go` might be a better place for this.
https://github.com/grafana/loki/blob/731ea38ac6b8d377c88858a4f45b5450aa0b8c0d/pkg/engine/planner/physical/optimizer.go#L140

For metric queries, we can apply the grouping columns columns from `RangeAggregation` as the requested keys.